### PR TITLE
DOC: try jupyterlite-xeus-python

### PR DESCRIPTION
**Pull request by benbovy**

### CODE REVIEW

**Comment by benbovy** [comment]
@martinRenou the build looks good, however I get a `semver range ^0.4.1 is not registered as a widget module` error in the JS console for ipytone widgets. I suspect it is related to installing ipytone in the doc environment in editable / development mode: https://github.com/benbovy/ipytone/blob/98f4cb905bf232aa2f1de233697fc58334a6a886/docs/environment.yml#L14-L18

I've tried disabling the `federated_extensions` jupyterlite add-on with `ignore_sys_prefix`, but then the xeus-python kernel was not available for some reason.

I've also tried installing ipytone's stable version in the doc environment (like you did with ipycanvas) and it worked. However, I find it convenient to install it from the current (main) branch for pull-request (latest) doc builds. It's actually required to execute the myst-nb user guide documents without error.

Is it possible to install ipytone from source such that it only installs the Python package without building and installing the federated extension? That's not very elegant but I think it would allow executing the myst-nb while avoiding conflicts in the jupyterlite xeus deployment. Better would be that jupyterlite avoid coping the ipytone federated extension from the doc environment.

**Comment by martinRenou** [comment]
I'll give your PR a try locally. Looking at the logs I don't see anything wrong, this is odd.

**Comment by benbovy** [comment]
> Are you sure you are using jupyterlite-xeus-python 0.4.2 (latest)?

We don't see in RTD logs the output of pip installed packages when creating the conda environment, but I assume it is the case in the [last build](https://readthedocs.org/projects/ipytone/builds/17245228/). I confirm that with this build there's no visible kernel in jupyterlite.

I've also tried locally after upgrading `jupyterlite`, `jupyterlite-sphinx` and `jupyterlite-xeus-python` to the versions below. I haven't seen anything wrong either in the logs, the xeus-python kernel is visible, but I have an unhelpful `uncaught exception: Object` error in the JS console (after `make_xkernel` messages) while trying to execute a code cell. My local env is big and messy, though, I'd really need to try in a clean env.

```
ipytone                   0.4.1                     dev_0    <develop>
jupyterlab                3.4.2              pyhd8ed1ab_0    conda-forge
jupyterlite               0.1.0b9                  pypi_0    pypi
jupyterlite-sphinx        0.4.9                    pypi_0    pypi
jupyterlite-xeus-python   0.4.2                    pypi_0    pypi
python                    3.8.13          h394c593_0_cpython    conda-forge
sphinx                    5.0.2              pyh6c4a22f_0    conda-forge
```

**Comment by martinRenou** [comment]
It looks to me that `"ignore_sys_prefix": ["federated_extensions"]` is simply discarding any federated extension that is added to the lite output, so we probably don't want this for now. 

When removing this option I see it working nicely locally 💯

**Comment by benbovy** [comment]
Hmm, after a few tries (updading mamba, clear package cache, create clean env) I still have the error in the JS console (Chrome 102.0.5005.115) and  `uncaught exception: Object` (Firefox):

```
Uncaught (in promise) WebAssembly.Exception: wasm exception
    at http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.wasm:wasm-function[15013]:0x6315ea
    at http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.wasm:wasm-function[3457]:0x2a348a
    at http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.wasm:wasm-function[3427]:0x29c9db
    at http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.wasm:wasm-function[3467]:0x2a44cc
    at http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.wasm:wasm-function[10536]:0x56d4d5
    at http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.wasm:wasm-function[10534]:0x56cbb5
    at http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.wasm:wasm-function[277]:0x112e92
    at http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.wasm:wasm-function[270]:0x11270a
    at xkernel.constructor$xkernel (eval at new_ (http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.js:1:1), <anonymous>:7:10)
    at xkernel.<anonymous> (http://localhost:8000/lite/extensions/@jupyterlite/xeus-python-kernel/static/xpython_wasm.js:9:174566)
```

I'm on MacOS, though. Maybe the lines below (404 failed) in the logs could explain it?

```
https://repo.mamba.pm/emscripten-forge/osx-64       91.0 B @ 144.0 B/s 404 failed  0.6s
Transaction

  Prefix: /tmp/xeus-python-kernel/envs/xeus-python-kernel
...
pkgs/main/emscripten-32                            285.0 B @   1.5kB/s 404 failed  0.2s
pkgs/r/emscripten-32                               284.0 B @   1.5kB/s 404 failed  0.2s
conda-forge/emscripten-32                          356.0 B @   1.7kB/s 404 failed  0.2s
https://repo.mamba.pm/conda-forge/emscripten-32     93.0 B @  27.0 B/s 404 failed  3.3s
Transaction

  Prefix: /tmp/xeus-python-kernel/envs/xeus-python-kernel

```

> When removing this option I see it working nicely locally 💯

It's weird that it doesn't work on RTD, then... (i.e., the `semver` error).

**Comment by benbovy** [comment]
Sorry, I've triggered the build again (https://readthedocs.org/projects/ipytone/builds/17296494/) and the ipytone extension is still missing. I get a 404 error for this resource: `https://ipytone--78.org.readthedocs.build/en/78/lite/extensions/ipytone/static/remoteEntry.fca5c8f66ba7b0140264.js`.

Unfortunately my local build still fails too after upgrading jupyterlite-xeus-python (cf. wasm exception above, but here the ipytone extension is correctly loaded in the front-end). Maybe I should try in a Linux VM (cf. emscripten-forge/osx-64 404  above)... I still need to open an issue on the xeus-python-kernel repo.

Happy to try it more. Thanks for your work on this (and also on jupyterlite-sphinx -- waiting for the next release to try the `jupyterlite_contents` config and see if I can include the user guide notebooks generated by myst-nb in the jupyterlite deployment).

File: docs/environment.yml [hunk 1]
```diff
@@ -10,9 +10,11 @@ dependencies:
   - matplotlib-base
   - myst-nb
   - yarn
+  - empack
```
Comment (martinRenou) [inline docs/environment.yml:13]: This should not be needed anymore

**Comment by martinRenou** [comment]
I believe the issue is that you are dev-building ipytone in the docs environment while you install a released version of ipytone in the emscripten environment. Both installed labextensions don't have the same hash for the index (`remoteEntry.fca5c8f66ba7b0140264.js`) and jupyterlite fails to pick the right extension entry (the one installed in the emscripten env, which overwrites the dev one) when deduplicating the extensions in the `jupyter-lite.json` config (see https://github.com/jupyterlite/jupyterlite/blob/main/py/jupyterlite/src/jupyterlite/addons/base.py#L261, it's not verifying the remote entry is correct when doing the dedupe).

I will push a quick fix to `jupyterlite-xeus-python` and another one upstream in jupyterlite. 

cc. @jtpio

**Comment by benbovy** [comment]
> I believe the issue is that you are dev-building ipytone in the docs environment while you install a released version of ipytone in the emscripten environment.

Yes, ideally I'd dev-build ipytone in the emscripten environment too so that the version in the jupyterlite deployment is always in sync with the version used for the corresponding doc build. But it might not be easy to provide such option in `jupyter-xeus-python`? Having all dev-dependencies (jupyterlab packaging, etc.) packaged in the emscripten environment is not ideal either I guess, unless it is possible to select packages to keep or discard when packaging the environment (I have zero knowledge on how this works).

**Comment by martinRenou** [comment]
Yes, it would be nice if one could simply use a dev install, at least for pure Python packages. We should definitely add a jupyterlite-xeus-python option for this.

**Comment by jtpio** [comment]
> it would be nice if one could simply use a dev install, at least for pure Python packages.

Agreed :+1: Many packages (and widgets) are indeed pure Python.

Maybe support for relative paths in `XeusPythonEnv.packages`?

**Comment by benbovy** [comment]
Will do as soon as 0.5.1 is published on pypi :-)

**Comment by benbovy** [comment]
Yes it works!! Thanks @martinRenou!

**Pull request merged**